src/objective-c/GRPCClient/GRPCCall.m

```diff
@@ -52,6 +55,165 @@ @interface GRPCCall ()<GRXWriteable>
+                                      callSafety:self->_requestOptions.safety
+                                  requestsWriter:self->_pipe
+                                     callOptions:self->_callOptions];
+    if (self->_callOptions.initialMetadata) {
+      [self->_call.requestHeaders addEntriesFromDictionary:self->_callOptions.initialMetadata];
+    }
+    id<GRXWriteable> responseWriteable = [[GRXWriteable alloc] initWithValueHandler:^(id value) {
+      dispatch_async(self->_dispatchQueue, ^{
+        if (self->_handler) {
+          id<GRPCResponseHandler> handler = self->_handler;
+          NSDictionary *headers = nil;
+          if (!self->_initialMetadataPublished) {
+            headers = self->_call.responseHeaders;
+            self->_initialMetadataPublished = YES;
+          }
+          if (headers) {
+            dispatch_async(handler.dispatchQueue, ^{
+              [handler receivedInitialMetadata:headers];
+            });
+          }
+          if (value) {
+            dispatch_async(handler.dispatchQueue, ^{
+              [handler receivedMessage:value];
+            });
+          }
+        }
+      });
+    }
+        completionHandler:^(NSError *errorOrNil) {
+          dispatch_async(self->_dispatchQueue, ^{
+            if (self->_handler) {
+              id<GRPCResponseHandler> handler = self->_handler;
+              NSDictionary *headers = nil;
+              if (!self->_initialMetadataPublished) {
+                headers = self->_call.responseHeaders;
+                self->_initialMetadataPublished = YES;
+              }
+              if (headers) {
+                dispatch_async(handler.dispatchQueue, ^{
+                  [handler receivedInitialMetadata:headers];
+                });
+              }
+              dispatch_async(handler.dispatchQueue, ^{
+                [handler closedWithTrailingMetadata:self->_call.responseTrailers error:errorOrNil];
+              });
+            }
+          });
+        }];
+    [self->_call startWithWriteable:responseWriteable];
+  });
+}
+
+- (void)cancel {
+  dispatch_async(_dispatchQueue, ^{
+    if (self->_call) {
+      [self->_call cancel];
+      self->_call = nil;
+    }
+    if (self->_handler) {
+      id<GRPCResponseHandler> handler = self->_handler;
+      dispatch_async(handler.dispatchQueue, ^{
+        [handler closedWithTrailingMetadata:nil
+                                      error:[NSError errorWithDomain:kGRPCErrorDomain
+                                                                code:GRPCErrorCodeCancelled
+                                                            userInfo:@{
+                                                              NSLocalizedDescriptionKey :
+                                                                  @"Canceled by app"
+                                                            }]];
+      });
+      self->_handler = nil;
```

Comment:
Probably worth a comment that by clearing handler here, you're ensuring that callbacks from GRXWriteable above are not run as well.

Reply:
The purpose of clearing handler here is to prevent more responses being issued to the handler (since no response should be issued after `closedWithTrailingMetadata`). I'll comment on the code.